aptos-move/move-examples/groth16_example/sources/groth16.move

```diff
@@ -0,0 +1,123 @@
+/// Generic implementation of Groth16 (proof verification) as defined in https://eprint.iacr.org/2016/260.pdf, Section 3.2.
+/// Actual proof verifiers can be constructed using the pairings supported in the generic algebra module.
+/// See the test cases in this module for an example of constructing with BLS12-381 curves.
+module groth16_example::groth16 {
+    use aptos_std::algebra::{Element, from_u64, multi_scalar_mul, eq, multi_pairing, upcast, pairing, add, zero};
+
+    /// Proof verification as specified in the original paper,
+    /// with the following input (in the original paper notations).
+    /// - Verification key: $\left([\alpha]_1, [\beta]_2, [\gamma]_2, [\delta]_2, \left\\{ \left[ \frac{\beta \cdot u_i(x) + \alpha \cdot v_i(x) + w_i(x)}{\gamma} \right]_1 \right\\}\_{i=0}^l \right)$.
+    /// - Public inputs: $\\{a_i\\}_{i=1}^l$.
+    /// - Proof $\left( \left[ A \right]_1, \left[ B \right]_2, \left[ C \right]_1 \right)$.
+    public fun verify_proof<G1,G2,Gt,S>(
+        vk_alpha_g1: &Element<G1>,
+        vk_beta_g2: &Element<G2>,
+        vk_gamma_g2: &Element<G2>,
+        vk_delta_g2: &Element<G2>,
+        vk_uvw_gamma_g1: &vector<Element<G1>>,
+        public_inputs: &vector<Element<S>>,
+        proof_a: &Element<G1>,
+        proof_b: &Element<G2>,
+        proof_c: &Element<G1>,
+    ): bool {
+        let left = pairing<G1,G2,Gt>(proof_a, proof_b);
+        let scalars = vector[from_u64<S>(1)];
+        std::vector::append(&mut scalars, *public_inputs);
+        let right = zero<Gt>();
+        let right = add(&right, &pairing<G1,G2,Gt>(vk_alpha_g1, vk_beta_g2));
+        let right = add(&right, &pairing(&multi_scalar_mul(vk_uvw_gamma_g1, &scalars), vk_gamma_g2));
+        let right = add(&right, &pairing(proof_c, vk_delta_g2));
+        eq(&left, &right)
+    }
+
+    /// Proof verification optimized for low verification latency but requiring pre-computation,
+    /// with the following input (in the original paper notations).
+    /// - Prepared verification key: $\left([\alpha]_1 \cdot [\beta]_2, -[\gamma]_2, -[\delta]_2, \left\\{ \left[ \frac{\beta \cdot u_i(x) + \alpha \cdot v_i(x) + w_i(x)}{\gamma} \right]_1 \right\\}\_{i=0}^l \right)$.
+    /// - Public inputs: $\\{a_i\\}_{i=1}^l$.
+    /// - Proof: $\left( \left[ A \right]_1, \left[ B \right]_2, \left[ C \right]_1 \right)$.
+    public fun verify_proof_prepared<G1,G2,Gt,GtParent,S>(
+        pvk_alpha_g1_beta_g2: &Element<GtParent>,
```

Comment:
I forgot, why can't this be a `Gt`?

Reply:
Because Gt deserializing + membership check is very expensive.

Reply:
But now it looks like we can save a bit more by:
- taking `pvk_alpha_g1_beta_g2_serialized: vector<u8>` instead of `pvk_alpha_g1_beta_g2: &Element<GtParent>` and
- instead Fq12 equality check, checking if `pvk_alpha_g1_beta_g2_serialized==serialize(...)`.

Reply:
don't use GtParent in `verify_proof_prepared`; make a `verify_proof_prepared_optimized` that use GtParent and explain the diff.

Reply:
changed to `verify_proof()` + `verify_proof_prepared()` + `verify_proof_prepared_fq12()`.

`verify_proof_prepared_optimized` can be confusing as there is nothing optimized inside the verification process.